File: tools/Utilities/BuildUtilities/Tasks/IncrementVersion.cs
namespace BuildUtilities
{
    using System;
    using System.Collections.Generic;
    using System.IO;
    using Microsoft.Build.Framework;
    using Microsoft.Build.Utilities;
    using ReplaceTokensLib;

    public static partial class ICBuildLib
    {
        public class IncrementVersion : AppDomainIsolatedTask
        {
            [Required]
            public string TFS_User { get; set; }

            [Required]
            public string TFS_Pass { get; set; }

            [Required]
            public string VersionXmlFile { get; set; }

            [Required]
            public string VersionCsFile { get; set; }

            [Required]
            public string ProjectRoot { get; set; }

            [Required]
            public string ProgramFiles { get; set; }

            [Required]
            public string TFExePath { get; set; }

            public string Version { get; set; }

            private string configFile = null;
            private string projectName = null;
            private string newVersion = null;

            public override bool Execute()
            {
                try
                {
                    Initialize(base.Log, this.ProgramFiles, this.TFExePath);
                    checkoutVersionFiles();
                    incrementVersion();
                    checkinVersionfiles();
                }
                catch (Exception ex)
                {
                    Log.LogErrorFromException(ex, true, true, null);
                    return false;
                }
                return true;
            }

            private void checkoutVersionFiles()
            {
                if (string.IsNullOrEmpty(Version))
                    TFSUtils.tfsGet(VersionXmlFile, TFS_User, TFS_Pass);
                TFSUtils.tfsGet(VersionCsFile, TFS_User, TFS_Pass);
                if (string.IsNullOrEmpty(Version))
                    TFSUtils.tfsCheckout(VersionXmlFile, TFS_User, TFS_Pass);
                TFSUtils.tfsCheckout(VersionCsFile, TFS_User, TFS_Pass);

                // TODO: Remove Telenor-specific logic
                // For Telenor builds only, check out config files
                if (VersionXmlFile.Contains("Telenor"))
                {
                    configFile = ProjectRoot + @"\Applications\WindowsService\App.config";
                    TFSUtils.tfsGet(configFile, TFS_User, TFS_Pass);
                    TFSUtils.tfsCheckout(configFile, TFS_User, TFS_Pass);
                }
            }

            private void checkinVersionfiles()
            {
                string tfscheckincomment;
                if (string.IsNullOrEmpty(Version))
                {
                    tfscheckincomment =
                        string.Format("IncrementVersion script: Incremented {0} IC version to {1}", projectName,
                                      newVersion);
                    TFSUtils.tfsCheckin(VersionXmlFile, tfscheckincomment, TFS_User, TFS_Pass);
                }

                tfscheckincomment = string.Format("IncrementVersion script: Updated {0} IC version.cs to {1}", projectName, newVersion);
                TFSUtils.tfsCheckin(VersionCsFile, tfscheckincomment, TFS_User, TFS_Pass);

                // TODO: Remove Telenor-specific logic
                // For Telenor builds only, check in config files
                if (VersionXmlFile.Contains("Telenor"))
                {
                    tfscheckincomment = string.Format("IncrementVersion script: Updated {0} {1} to {2}", projectName, configFile.Substring(configFile.LastIndexOf('\\') + 1), newVersion);
                    TFSUtils.tfsCheckin(configFile, tfscheckincomment, TFS_User, TFS_Pass);
                }
            }

            private void incrementVersion()
            {
                if (!File.Exists(VersionXmlFile))
                    throw new Exception(string.Format("Unable to find version xml file '{0}'.", VersionXmlFile));

                projectName = FileUtils.GetValueFromVersionFile(VersionXmlFile, "//ProjectName");
                var incrementText = FileUtils.GetValueFromVersionFile(VersionXmlFile, "//Increment");
                var lastVersionText = FileUtils.GetValueFromVersionFile(VersionXmlFile, "//LastVersion");
                if (!string.IsNullOrEmpty(Version))
                    lastVersionText = Version;
                var versionElements = lastVersionText.Split(new char[] { '.' });
                var newmajor = int.Parse(versionElements[0]);
                var newminor = int.Parse(versionElements[1]);
                var newrevision = int.Parse(versionElements[2]);
                var newbuild = int.Parse(versionElements[3]);

                if (string.IsNullOrEmpty(Version))
                {
                    switch (incrementText.ToLower())
                    {
                        case "major":
                            newmajor++;
                            break;
                        case "minor":
                            newminor++;
                            break;
                        case "revision":
                            newrevision++;
                            break;
                        case "build":
                            newbuild++;
                            break;
                    }
                }

                newVersion = string.Format("{0}.{1}.{2}.{3}", newmajor, newminor, newrevision, newbuild);
                if (string.IsNullOrEmpty(Version))
                    FileUtils.UpdateValueInVersionFile(VersionXmlFile, "//LastVersion", newVersion);

                var lines = File.ReadAllLines(VersionCsFile);
                string newAssemblyFileVersion = null;

                for (var i = 0; i < lines.Length; i++)
                {
                    if (lines[i].StartsWith("[assembly: AssemblyVersion"))
                    {
                        string newAssemblyVersion;
                        string assemblyVersion;
                        GetAssemblyVersionInfo(lines[i], newmajor, newminor, newrevision, newbuild, out assemblyVersion, out newAssemblyVersion);
                        lines[i] = lines[i].Replace(assemblyVersion, newAssemblyVersion);
                    }
                    else if (lines[i].StartsWith("[assembly: AssemblyFileVersion"))
                    {
                        string assemblyFileVersion;
                        GetAssemblyVersionInfo(lines[i], newmajor, newminor, newrevision, newbuild, out assemblyFileVersion, out newAssemblyFileVersion);
                        lines[i] = lines[i].Replace(assemblyFileVersion, newAssemblyFileVersion);
                    }
                }

                File.WriteAllLines(VersionCsFile, lines);

                // TODO: Remove Telenor-specific logic
                // If Telenor build, update config file
                if (VersionXmlFile.Contains("Telenor"))
                {
                    var template = configFile + ".template";
                    var tokens = new Dictionary<string, string>();
                    tokens.Add("@@VERSION@@", newAssemblyFileVersion);
                    TokenUtils.ReplaceTokens(template, configFile, tokens);
                }
            }

            private void GetAssemblyVersionInfo(string line, int newmajor, int newminor, int newrevision, int newbuild, out string assemblyVersion, out string newAssemblyVersion)
            {
                assemblyVersion = line.Split('"')[1];
                var assemblyVersionMajor = assemblyVersion.Split('.')[0];
                var assemblyVersionMinor = assemblyVersion.Split('.')[1];
                var assemblyVersionRevision = assemblyVersion.Split('.')[2];
                var assemblyVersionBuild = assemblyVersion.Split('.')[3];
                if (assemblyVersionMajor != "*")
                    assemblyVersionMajor = newmajor.ToString();
                if (assemblyVersionMinor != "*")
                    assemblyVersionMinor = newminor.ToString();
                if (assemblyVersionRevision != "*")
                    assemblyVersionRevision = newrevision.ToString();
                if (assemblyVersionBuild != "*")
                    assemblyVersionBuild = newbuild.ToString();
                newAssemblyVersion = string.Format("{0}.{1}.{2}.{3}", assemblyVersionMajor, assemblyVersionMinor, assemblyVersionRevision, assemblyVersionBuild);
            }
        }
    }
}

File: tools/scripts/Python/3.2/incrementversion.py
import subprocess, os, sys
import glob, ibslib
from optparse import OptionParser
from shutil import *
from xml.etree.ElementTree import Element, SubElement, ElementTree

class IncrementVersion():
    def __init__(self):
        self.main()

    def showUsage(self):
        print("\
Usage: incrementversion.py [options]\n\n\
Options:\n\
    -h, --help            show this help message and exit\n\
    --tfs-user=TFS_USER\n\
    --tfs-pass=TFS_PASS\n\
    --versionxmlfile=(path to version xml file)\n\
    --versioncsfile=(path to version cs file)\n\
    --projectroot=(path of project)\n")
        exit(1)

    def checkoutVersionFiles(self):
        ibslib.tfsGet(self.options.versionxmlfile, self.options.tfs_user, self.options.tfs_pass)
        ibslib.tfsGet(self.options.versioncsfile, self.options.tfs_user, self.options.tfs_pass)
        ibslib.tfsCheckout(self.options.versionxmlfile, self.options.tfs_user, self.options.tfs_pass)
        ibslib.tfsCheckout(self.options.versioncsfile, self.options.tfs_user, self.options.tfs_pass)
        
        # For Telenor builds only, check out config files
        if self.options.versionxmlfile.__contains__('Telenor'):
            ibslib.tfsGet(self.configfile, self.options.tfs_user, self.options.tfs_pass)
            ibslib.tfsCheckout(self.configfile, self.options.tfs_user, self.options.tfs_pass)

    def checkinVersionFiles(self):
        tfscheckincomment = "IncrementVersion script: Incremented " + self.projectname + " IC version to " + self.newversion
        ibslib.tfsCheckin(self.options.versionxmlfile, tfscheckincomment, self.options.tfs_user, self.options.tfs_pass)

        tfscheckincomment = "IncrementVersion script: Updated " + self.projectname + " IC version.cs to " + self.newversion
        ibslib.tfsCheckin(self.options.versioncsfile, tfscheckincomment, self.options.tfs_user, self.options.tfs_pass)

        # For Telenor builds only, check in config files
        if self.options.versionxmlfile.__contains__('Telenor'):
            tfscheckincomment = "IncrementVersion script: Updated " + self.projectname + " " + self.configfile.split("\\")[len(self.configfile.split("\\"))-1] + " to " + self.newversion
            ibslib.tfsCheckin(self.configfile, tfscheckincomment, self.options.tfs_user, self.options.tfs_pass)

    def incrementVersion(self):
        if os.path.exists(self.options.versionxmlfile):
            tree = ElementTree(file=self.options.versionxmlfile)
            IncrementNode = tree.find(".//Increment")
            if IncrementNode == None:
                  print("Increment element not found in version XML file!")
                  exit(1)
            incrementText = "".join(IncrementNode.itertext())
            LastVersionNode = tree.find(".//LastVersion")
            if LastVersionNode == None:
                  print("LastVersion element not found in version XML file!")
                  exit(1)
            lastVersionText = "".join(LastVersionNode.itertext())
            ProjectNameNode = tree.find(".//ProjectName")
            if ProjectNameNode == None:
                  print("ProjectName element not found in version XML file!")
                  exit(1)
            self.projectname = "".join(ProjectNameNode.itertext())

            newmajor = int(lastVersionText.split('.')[0])
            newminor = int(lastVersionText.split('.')[1])
            newrevision = int(lastVersionText.split('.')[2])
            newbuild = int(lastVersionText.split('.')[3])
            
            if incrementText.lower() == "major":
                newmajor = str(eval("newmajor+1"))
            elif incrementText.lower() == "minor":
                newminor = str(eval("newminor+1"))
            elif incrementText.lower() == "revision":
                newrevision = str(eval("newrevision+1"))
            elif incrementText.lower() == "build":
                newbuild = str(eval("newbuild+1"))
                
            newversion = str(newmajor) + "." + str(newminor) + "." + str(newrevision) + "." + str(newbuild)

            self.newversion = newversion
            LastVersionNode.text = newversion
            tree.write(self.options.versionxmlfile)

            fi = open(self.options.versioncsfile, "r")
            lines = fi.readlines()
            outlines = []
            for line in lines:
                if line.startswith("[assembly: AssemblyVersion"):
                    assemblyVersion = line.split('"')[1]
                    assemblyVersionMajor = assemblyVersion.split('.')[0]
                    assemblyVersionMinor = assemblyVersion.split('.')[1]
                    assemblyVersionRevision = assemblyVersion.split('.')[2]
                    assemblyVersionBuild = assemblyVersion.split('.')[3]
                    if (assemblyVersionMajor != "*"):
                        assemblyVersionMajor = str(newmajor)
                    if (assemblyVersionMinor != "*"):
                        assemblyVersionMinor = str(newminor)
                    if (assemblyVersionRevision != "*"):
                        assemblyVersionRevision = str(newrevision)
                    if (assemblyVersionBuild != "*"):
                        assemblyVersionBuild = str(newbuild)
                    newAssemblyVersion = assemblyVersionMajor + "." + assemblyVersionMinor + "." + assemblyVersionRevision + "." + assemblyVersionBuild
                    outlines.append(line.replace(assemblyVersion, newAssemblyVersion))
                elif line.startswith("[assembly: AssemblyFileVersion"):
                    assemblyFileVersion = line.split('"')[1]
                    assemblyFileVersionMajor = assemblyFileVersion.split('.')[0]
                    assemblyFileVersionMinor = assemblyFileVersion.split('.')[1]
                    assemblyFileVersionRevision = assemblyFileVersion.split('.')[2]
                    assemblyFileVersionBuild = assemblyFileVersion.split('.')[3]
                    if (assemblyFileVersionMajor != "*"):
                        assemblyFileVersionMajor = str(newmajor)
                    if (assemblyFileVersionMinor != "*"):
                        assemblyFileVersionMinor = str(newminor)
                    if (assemblyFileVersionRevision != "*"):
                        assemblyFileVersionRevision = str(newrevision)
                    if (assemblyFileVersionBuild != "*"):
                        assemblyFileVersionBuild = str(newbuild)
                    newassemblyFileVersion = assemblyFileVersionMajor + "." + assemblyFileVersionMinor + "." + assemblyFileVersionRevision + "." + assemblyFileVersionBuild
                    outlines.append(line.replace(assemblyFileVersion, newassemblyFileVersion))
                else:
                    outlines.append(line)
            fi.close()
            fi = open(self.options.versioncsfile, "w")
            for line in outlines:
                fi.write(line)
            fi.close()

            # If Telenor build, update config file
            if self.options.versionxmlfile.__contains__('Telenor'):
                configfile = self.configfile
                template = configfile + '.template'
                if os.path.exists(configfile):
                    os.remove(configfile)
                fi = open(template, "r")
                fo = open(configfile, "w")
                for line in fi.readlines():
                    fo.write(line.replace("@@VERSION@@", newassemblyFileVersion))
                fo.close()
                fi.close()
        else:
            print("Version XML file not found!")
            exit(1)

    def main(self):
        parser = OptionParser()
        parser.add_option("--versionxmlfile", action="store", type="string", dest="versionxmlfile")
        parser.add_option("--versioncsfile", action="store", type="string", dest="versioncsfile")
        parser.add_option("--tfs-user", action="store", type="string", dest="tfs_user")
        parser.add_option("--tfs-pass", action="store", type="string", dest="tfs_pass")
        parser.add_option("--projectroot", action="store", type="string", dest="projectroot")
        (options, args) = parser.parse_args()
        if (self.validateOptions(options) == False):
            self.showUsage
        else:
            self.options = options
        if self.options.versionxmlfile.__contains__('Telenor'):
            self.configfile = self.options.projectroot + "\\Applications\\WindowsService\\App.config"
        self.checkoutVersionFiles()
        self.incrementVersion()
        self.checkinVersionFiles()

    def validateOptions(self, options):
        if options.versionxmlfile == None or \
            options.versioncsfile == None or \
            options.tfs_user == None or \
            options.tfs_pass == None:
            return False
        else:
            return True

IncrementVersion()
